### PR TITLE
[next] Closes #7066: Add support for parent-child device types to enable multi-level nested devices

### LINKED ISSUE #7066: RFE: Multiply nestable Device objects
### As ...

Dan - Data Owner

### I want ...

To be able to nest `Device` objects more than one level in the same way that `Module` objects can be nested today.

### So that ...

I can model devices like the [Supermicro ARS-121L-DNR](https://www.supermicro.com/en/products/system/gpu/1u/ars-121l-dnr). These devices ship as a single monolithic device with a single serial number on the chassis. They consist of two half-width servers in an enclosure. The enclosure has no CPU; its essentially just a power bus for the the two servers which can be racked. Each server is an entirely independent computer which boots an OS image. In our case, each server contains a Bluefield PCI card which /also/ boots its own OS image. Thus, we have something like:

```
Enclosure (Device )
  |
  |
   -> Node A - half-width server (Device)
  |      |
  |      |
  |       -> BlueField card (Device)
  |
   -> Node B - half-width server (Device)
         |
         |
          -> BlueField card (Device)
```

Since both the Nodes and the Bluefield cards boot OS images, I'd like to be able to model them both as `Device` objects.

### I know this is done when...

I can create `Device` object with a multiply-nested collection of child/grandchild (optionally even deeper) `Device` objects.

### Optional - Feature groups this request pertains to.

- [ ] Automation
- [ ] Circuits
- [x] DCIM
- [ ] IPAM
- [ ] Misc (including Data Sources)
- [ ] Organization
- [ ] Apps (and other Extensibility)
- [ ] Security (Secrets, etc)
- [ ] Image Management
- [ ] UI/UX
- [ ] Documentation
- [ ] Other (not directly a platform feature)

### Database Changes

It would almost certainly require changes of some sort to the data models. Don't know the specifics, tho.

### External Dependencies

I don't believe so.

### PR DESCRIPTION
**Pull request by nrnvgh**
# Closes #7066 
# What's Changed

This is a version of #8592 suitable to merging into `next`.

Added `subdevice_role` called `parent-child` which can function as both a Parent and Child `Device`. This enables nested `Device` objects to arbitrary depths. There's not currently a hard limitation on object depth, although `Device` does currently have the same soft limit as `Module`, where there's a depth limit of 4 on queries, but you can build topologies deeper than than that.

The screenshots below show the modeling of a 3-level system:

1. Chassis (Parent), with 28 bays
2. 2-node Sled in each Chassis Bay (Parent-Child)
3. 1 motherboard node in each of the Sled bays

The current patch applies cleanly to `next` as of 2026/02/17 @ 22:00 GMT.

## Nested saves

Nested saves work by querying the immediate children of the `Device` being saved and (as with the current release of the code) if either `rack` or `location` has changed on the parent `Device`, the children are saved. If the parent's bays have devices which are `parent-child`, the same logic applies, each one will have its children fetched and the rack/location will be updated if needed.

# Screenshots

## DeviceType - Before
Screenshot 2026-02-17 at 3 32 37 PM

## DeviceType - After
Screenshot 2026-02-17 at 3 34 02 PM

#### Device Type
Screenshot 2026-02-17 at 3 37 14 PM

#### Device
Screenshot 2026-02-17 at 3 36 14 PM

Screenshot 2026-02-17 at 3 36 27 PM

#### Device Type
Screenshot 2026-02-17 at 3 37 32 PM

#### Device

Screenshot 2026-02-17 at 3 36 40 PM

Screenshot 2026-02-17 at 3 36 47 PM

#### Device Type
Screenshot 2026-02-17 at 3 37 25 PM

#### Device

Screenshot 2026-02-17 at 3 36 56 PM

### Query

```graphql

fragment DeviceDetails on DeviceType {
  name
  device_type {
    model
    manufacturer {
      name
    }
  }
}

query DeviceTree {
  devices(name: "sm-chassis-01") {
    ...DeviceDetails
    device_bays {
      name
      installed_device {
        ...DeviceDetails
        device_bays {
          installed_device {
            ...DeviceDetails
          }
        }
      }
    }
  }
}
```

### Response (partial)
```json
{
  "data": {
    "devices": [
      {
        "name": "sm-chassis-01",
        "device_type": {
          "model": "Supermicro Test Chassis 6U",
          "manufacturer": {
            "name": "Supermicro"
          }
        },
        "device_bays": [
          {
            "name": "Bay01",
            "installed_device": {
              "name": "sm-chassis-01-blade01",
              "device_type": {
                "model": "Supermicro Test Blade",
                "manufacturer": {
                  "name": "Supermicro"
                }
              },
              "device_bays": [
                {
                  "installed_device": {
                    "name": "sm-chassis-01-blade01-node01",
                    "device_type": {
                      "model": "Supermicro Test Motherboard Node",
                      "manufacturer": {
                        "name": "Supermicro"
                      }
                    }
                  }
                },
                {
                  "installed_device": {
                    "name": "sm-chassis-01-blade01-node02",
                    "device_type": {
                      "model": "Supermicro Test Motherboard Node",
                      "manufacturer": {
                        "name": "Supermicro"
                      }
                    }
                  }
                }
              ]
            }
          },
    [...]
```

# TODO

- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/core/#creating-changelog-fragments))
- [x] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [x] Documentation Updates (when adding/changing features)
- [ ] Example App Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/dcim/views.py [hunk 1]
```diff
@@ -2540,14 +2540,35 @@ def render_value(self, key, value, context):
...
+
+                        while current_bay is not None and current_depth < max_depth:
+                            path_objects.insert(0, current_bay)
+                            current_device = current_bay.device
+                            if current_device.pk in visited_device_ids:
+                                break
+
+                            visited_device_ids.add(current_device.pk)
+                            path_objects.insert(0, current_device)
+                            current_bay = getattr(current_device, "parent_bay", None)
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/dcim/views.py]: This seems legitimate at a glance.

File: nautobot/dcim/models/device_components.py [hunk 2]
```diff
@@ -1165,21 +1164,51 @@ def clean(self):
...
-            if self.installed_device.device_type.subdevice_role != SubdeviceRoleChoices.ROLE_CHILD:
+            if parent_device in seen_devices:
                 raise ValidationError(
-                    {
-                        "installed_device": f'Cannot install device "{self.installed_device}"; device-type "{self.installed_device.device_type}" subdevice_role is not "child".'
-                    }
+                    "The device parent chain already contains a loop; fix existing data before making this assignment."
                 )
+            seen_devices.add(parent_device)
+            parent_bay = DeviceBay.objects.filter(installed_device=parent_device).first()
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/dcim/models/device_components.py]: Seems like a hallucination, although the proposed change isn't necessarily wrong it seems unnecessary:

```py
>>> x = set()
>>> x.add(Device.objects.first())
>>> x
{<Device: Access Point-17>}
>>> x.add(Device.objects.last())
>>> x
{<Device: WLAN Controller-30>, <Device: Access Point-17>}
>>> 
```

File: nautobot/dcim/tests/test_forms.py [hunk 3]
```diff
@@ -274,6 +277,110 @@ def test_interface_label_count_mismatch(self):
...
+            location=cls.location,
+            rack=cls.rack,
+            status=cls.status,
+        )
+
+    def test_installed_device_queryset_includes_parent_child_role(self):
+        form = PopulateDeviceBayForm(self.device_bay)
+        queryset = form.fields["installed_device"].queryset
+
+        self.assertTrue(queryset.filter(pk=self.child_device.pk).exists())
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_forms.py]: Possibly could use `assertIn(self.child_device, queryset)` etc. instead, but this is fine as well.

**Review by glennmatthews — APPROVED** [review]

File: nautobot/dcim/models/device_components.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nrnvgh) [reply nautobot/dcim/models/device_components.py]: I tend to agree. That said, storing instances will probably consume a bit more memory than just UUIDs. There are also some weird edge cases which wouldn't affect this particular block of code, but given that the UUID is all that's needed for this code, I don't see a downside do using the pk. FWIW, I spot-checked the core I saw /very/ few places where `set()` was used to create a collection of nautobot objects. I'll push an update for this shortly.

File: nautobot/dcim/tests/test_forms.py (hunk 3 above)
Comment (nrnvgh) [reply nautobot/dcim/tests/test_forms.py]: I've simplified this; one `assert` to rule them all, and so forth.

File: nautobot/dcim/views.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nrnvgh) [reply nautobot/dcim/views.py]: The current code works as expected, but I've refactored it to be a bit cleaner.

File: nautobot/dcim/views.py [hunk 4]
```diff
@@ -2535,43 +2535,63 @@ class DeviceFieldsPanel(object_detail.ObjectFieldsPanel):
...
+            while current_bay is not None and hop_count < max_hops:
+                current_device = current_bay.device
+                if current_device.pk in visited_device_ids:
+                    break
+
+                visited_device_ids.add(current_device.pk)
+                breadcrumb_segments.extend([current_bay, current_device])
+
+                current_bay = cls._get_parent_bay_or_none(current_device)
+                hop_count += 1
```
Comment (glennmatthews) [inline nautobot/dcim/views.py:2575]: Should this be `2` instead? I'm not clear on why we multiply by 2 in calculating `max_hops` but only increment `hop_count` by 1 each loop given that we're adding two elements to `breadcrumb_segments` each time.

Comment (nrnvgh) [reply nautobot/dcim/views.py:2575]: Sigh...yes. Apologies for that, it was a silly mistake, and thanks for catching it. I've re-worked that section and added both comments and test cases.

**Comment by nrnvgh** [comment]
Something I wanted to call out. I don't think it's a big deal as it's really an edge case, but did want you to be aware. `Device.all_nested_devices` builds an ORM query top-down whilst the breadcrumb builder builds the chain from the bottom up. This means that, in a very specific set of cases, they'll return different info. You'd need to have `DEVICE_RECURSION_DEPTH_LIMIT + 2` devices in a chain, which seems unlikely to say the least, but if you did then `all_nested_devices` would only return `DEVICE_RECURSION_DEPTH_LIMIT` devices after the queried device (e.g. device[1-4]; device[0] isn't used for that query) whilst the breadcrumb builder would have device[2-5]. As a for-instance:

```
devices=[
  <Device: breadcrumb-chain-device-6-0>,
  <Device: breadcrumb-chain-device-6-1>,
  <Device: breadcrumb-chain-device-6-2>,
  <Device: breadcrumb-chain-device-6-3>,
  <Device: breadcrumb-chain-device-6-4>,
  <Device: breadcrumb-chain-device-6-5>,
  <Device: breadcrumb-chain-device-6-6>
]

breadcrumb_objects=[
  <Device: breadcrumb-chain-device-6-2>,
  <DeviceBay: breadcrumb-chain-bay-6-2>,
  <Device: breadcrumb-chain-device-6-3>,
  <DeviceBay: breadcrumb-chain-bay-6-3>,
  <Device: breadcrumb-chain-device-6-4>,
  <DeviceBay: breadcrumb-chain-bay-6-4>,
  <Device: breadcrumb-chain-device-6-5>,
  <DeviceBay: breadcrumb-chain-bay-6-5>
]

devices[0].all_nested_devices=[
  <Device: breadcrumb-chain-device-6-1>,
  <Device: breadcrumb-chain-device-6-2>,
  <Device: breadcrumb-chain-device-6-3>,
  <Device: breadcrumb-chain-device-6-4>
]
```
